libraries/SITL/SIM_Scrimmage.cpp

```diff
@@ -0,0 +1,238 @@
+            tokens.push_back(intermediate);
+        }
+
+        // If the key:value was parsed correctly, add it to the config map
+        if (tokens.size() == 2) {
+            configs_[tokens[0]] = tokens[1];
+        }
+    }
+
+    // Were any of the defaults overwritten?
+    auto it_mission = configs_.find("mission");
+    if (it_mission != configs_.end()) {
+        mission_name_ = it_mission->second;
+    }
+
+    auto it_motion = configs_.find("motion_model");
+    if (it_motion != configs_.end()) {
+        motion_model_ = it_motion->second;
+    }
+
+    auto it_visual = configs_.find("visual_model");
+    if (it_visual != configs_.end()) {
+        visual_model_ = it_visual->second;
+    }
+
+    auto it_terrain = configs_.find("terrain");
+    if (it_terrain != configs_.end()) {
+        terrain_ = it_terrain->second;
+    }
+
+    // start scrimmage after parsing simulation configuration
+    start_scrimmage();
+}
+
+// Start Scrimmage child
+void Scrimmage::start_scrimmage(void)
+{
+    pid_t child_pid = fork();
+    if (child_pid == 0) {
+        // In child
+
+        // Construct the scrimmage command string with overrides for initial
+        // position and heading.
+        std::string full_exec_str = "'scrimmage " + mission_name_
+            + " -o \""
+            + "latitude_origin=" + std::to_string(home.lat / 1.0e7f) + ","
+            + "longitude_origin=" + std::to_string(home.lng / 1.0e7f) + ","
+            + "altitude_origin=" + std::to_string(home.alt / 1.0e2f) + ","
+            + "heading=" + std::to_string(home_yaw) + ","
+            + "motion_model=" + motion_model_ + ","
+            + "visual_model=" + visual_model_ + ","
+            + "terrain=" + terrain_ + ","
+            + "to_ardupilot_port=" + std::to_string(fdm_port+1) + ","
+            + "from_ardupilot_port=" + std::to_string(fdm_port)
+            + "\"'";
+
+        std::cout << full_exec_str.c_str() << std::endl;
+
+        // system call worked
+        full_exec_str = "xterm +hold -T SCRIMMAGE -e " + full_exec_str;
+        int ret = system(full_exec_str.c_str());
+
+        if (ret != 0) {
+            std::cerr << "scrimmage didn't open.\n";
+            perror("scrimmage");
+        }
+    }
+}
+
+/*2
+Serial port 2 on TC
+  send servos
+*/
+void Scrimmage::send_servos(const struct sitl_input &input)
+{
+    servo_packet pkt;
+
+    for (int i = 0; i < MAX_NUM_SERVOS; i++) {
+        pkt.servos[i] = input.servos[i];
+    }
+    send_sock.sendto(&pkt, sizeof(servo_packet), "127.0.0.1", fdm_port);
+}
+
+/*
+  receive an update from the FDM
+  This is a blocking function
+ */
+void Scrimmage::recv_fdm(const struct sitl_input &input)
+{
+    fdm_packet pkt;
+
+    // Re-send the servo packet every 0.1 seconds until we get a reply. This
+    // allows us to cope with some packet loss to the FDM
+    while (recv_sock.recv(&pkt, sizeof(pkt), 100) != sizeof(pkt)) {
+        //send_servos(input);
```

Comment:
what is this while() still doing here?

Reply:
This while loop waits for a packet of the correct size from scrimmage before continuing. Comment was updated to reflect this purpose.